bin/provision-instances

```diff
@@ -0,0 +1,227 @@
+#!/usr/bin/python3
+#
+# This script runs a full instance provision over all non-ephemeral instances
+# on OCIM. Running this script requires Django to be initialized - the easiest
+# way to do this is by evaluating the script from within the Django shell by
+# running:
+#
+#   > eval(open('bin/provision-instances').open())
+#
+# The provision process happens in "waves", where all instances that fail to
+# provision in a given wave are recorded and reprovisioned during the next wave.
+#
+# This provision script is designed to withstand restarts of either OCIM or
+# this script itself; it accomplishes this by always saving and loading the
+# state of the provision process to/from `ocim_update_waves.ini`.
+
+import ast
+from configparser import ConfigParser
+from datetime import datetime, timedelta
+import os
+import pytz
+from time import sleep
+
+from instance.models.openedx_instance import OpenEdXInstance
+from instance.tasks import spawn_appserver
+
+WAVES_FILE = 'ocim_update_waves.ini'
+SLEEP_TIME = 60
+MAX_PROCESSING = 6
+MAX_PROCESSING_TIME = 240  # 4 hours
+
+waves_config = ConfigParser()
+
+
+def bootstrap_instance_list():
+    # Collect all of the instances that should be included in the OCIM upgrades.
+    #
+    # Instances with any of the following attributes at the start of the upgrade
+    # are avoided for the entire upgrade:
+    #   - Any instance considered "shut down" (all VMs terminated)
+    #   - Any instance without any appservers
+    #   - Any instance set to use ephemeral databases
+    #   - Any instance where the latest appserver isn't in a healthy stable state (e.g. running)
+
+    failed_results = []
+    ready_instances = []
+    for inst in OpenEdXInstance.objects.all():
+        # Set field defaults for ALL instances, even ones we're not going to reprovision.
+        inst.set_field_defaults()
+        inst.save()
+
+        latest_appserver = inst.appserver_set.last()
```

Comment:
OK - I would still want to refactor this logic with the one identifying RC servers, so we can consistently identify RC servers, and whether the instance is ready to upgrade. It will also be helpful from the UI end to have this standardized.